### DISCUSSION TITLE
Is it possible to force Nitro to run in "regular" Node.js server mode when in dev mode?

### DISCUSSION THREAD

**Discussion by lirantal** [nuxt/nuxt#28050] [Questions]
I run my Nuxt 3 project using the regular `npm run dev` which seems to spin off Nitro inside a worker thread. I assume this is what's happening because one of my project dependencies uses `process.chdir()` which isn't allowed inside a Worker Thread, based on this error I received:

![image](https://github.com/nuxt/nuxt/assets/316371/a704dfdb-1e05-445e-b36a-56cc8e824cdd)

I assume this is due to the way Nitro works when I run `npm run dev` because I found this in the Nitro docs:

![image](https://github.com/nuxt/nuxt/assets/316371/f2ab432c-8220-4ec6-977f-2bd659fa5ca4)

To try and override this behavior I added the following to my `nuxt.config.ts`:

```json
$development: {
    nitro: {
      preset: 'node-server'
    }
  },
```

However, still getting the same worker thread error.

Would appreciate any advice.

**Comment by pi0** [comment]
We can certainly cover this by adding the support of a new dev server runner that forks a new process on each (re)load (while it means slower reloads) but it is on the roadmap as nitro will soon support running dev server in target isolates (like workerd for cloufdlare) 

I think you should be able to directly fix this issue by mocking `process.chdir = () => {}` in a nitro plugin or earlier in startup but i would help better if know exact library is causing issue.

**Reply by lirantal** [reply]
What would it be helpful to mock the chdir though? as in, I actually do want the library or my own code to be able to run that chdir command, not to disable it :)

**Reply by LouisCassany** [reply]
I'm exactly in the same case, trying to use [node serial port](https://serialport.io/) in a custom nitro plugin. When running `npm run dev`, the server crashes instantly because the library doesn't support worker environnements. Everything works fine on production but having to build everything during development is a bit tedious to say the least. @lirantal did you make any progress on this by any chance? :)

**Reply by pi0** [reply]
> actually do want the library or my own code to be able to run that chdir command,

This would affect other logic such as serve static. Trying to understand why exactly a process chdir is required. 

> I'm exactly in the same case, trying to use [node serial port](https://serialport.io/) in a custom nitro plugin. When running npm run dev, the server crashes instantly because the library doesn't support worker environments

Do you know why serial port library does not support node workers?

**Reply by LouisCassany** [reply]
> Do you know why serial port library does not support node workers?

Probably because of the underlying c++ code, as referenced [here](https://github.com/serialport/node-serialport/issues/1938#issuecomment-534951904).  I'm really glad its working in production in a `node-server` preset / crashing in dev and not the other way around but the dev experience is pretty bad, having to build the server and restart it every time you make a change.

**Reply by lirantal** [reply]
There's no actual workaround for the cases where you do need to use fs.chdir or other APIs that aren't supported in a worker environment. Would indeed wish there would be some way to at least configure this behavior, even if the impact is performance, it would be a conscious choice of the user.

**Reply by LouisCassany** [reply]
As a workaround, in my specific case, I wrote a quick native program that reads the serial port and output bytes to stdout. I launch it with a child process and then interact with it.

**Reply by lirantal** [reply]
Right, happy for you. My case uses a third-party dependency that does it. Forking or patching it is not something sustainable that I want to do :-)

**Comment by this-is-martin** [comment]
same problem here - I'd appreciate a fix :)

**Reply by pi0** [reply]
Have you tried workaround mentioned in https://github.com/nuxt/nuxt/discussions/28050#discussioncomment-10788885? If not working what is the affected library?
